mcstatus/protocol/connection.py

```diff
@@ -20,199 +22,501 @@
 
 
 def ip_type(address: Union[int, str]) -> Optional[int]:
+    """Returns what version of IP a given address is."""
     try:
         return ip_address(address).version
     except ValueError:
         return None
 
 
-class Connection:
-    def __init__(self):
-        self.sent = bytearray()
-        self.received = bytearray()
+class BaseWriteSync(ABC):
+    """Base synchronous write class"""
 
-    def read(self, length: int) -> bytearray:
-        result = self.received[:length]
-        self.received = self.received[length:]
-        return result
+    __slots__ = ()
 
+    @abstractmethod
     def write(self, data: Union["Connection", str, bytearray, bytes]) -> None:
-        if isinstance(data, Connection):
-            data = data.flush()
-        if isinstance(data, str):
-            data = bytearray(data, "utf-8")
-        self.sent.extend(data)
+        """Write data to self."""
+        ...
 
-    def receive(self, data: Union[BytesConvertable, bytearray]) -> None:
-        if not isinstance(data, bytearray):
-            data = bytearray(data)
-        self.received.extend(data)
+    def __repr__(self) -> str:
+        """Return representation of self."""
+        return f"<{self.__class__.__name__} Object>" ""
```

Comment:
There is never a "need" per say for a repr, but it's good practice to do it anyhow.

Reply:
And why use dataclasses when this will suffice? Plus I feel like there are problems with __slots__ in dataclasses in older versions or something, and if this is networking code faster attribute access is important.

Reply:
Yeah, slotted dataclasses are 3.10+, though you can jusr define `__slots__` explicitly, but this caused problems with inheritance iirc. I would agree that in this case dataclasses are an overkill.

Reply:
This is now tracked in #355, marking this discussion as resolved